src/Hal/JsonSchema/SchemaFactory.php

```diff
@@ -79,8 +78,15 @@ public function buildSchema(string $className, string $format = 'jsonhal', strin
             $schema['type'] = 'object';
             $schema['properties'] = [
                 '_embedded' => [
-                    'type' => 'array',
-                    'items' => $items,
+                    [
+                        'type' => 'object',
+                        'properties' => [
+                            'items' => [
+                                'type' => 'array',
+                                'items' => $items,
+                            ],
+                        ],
+                    ],
```

Comment:
something's bothering me here I need to check that, I don't understand why we put `array` in the first place.

Reply:
Introduced in #4357, as was `addDistinctFormat()`

Reply:
https://datatracker.ietf.org/doc/html/draft-kelly-json-hal#section-4.1.2 

> It is an object whose property names are link relation types (as defined by [[RFC5988](https://www.rfc-editor.org/info/rfc5988)]) and values are either a Resource Object or an array of Resource Objects.

can you revert this change? I think that our HAL representation always uses an array and we use `items` as key.

Reply:
Doesn't that make the behaviour in `ApiPlatform\Hal\Serializer\CollectionNormalizer` incorrect?

https://github.com/api-platform/core/blob/874e4d6707ecdd48b27ed91b0b66ea5e33691e7c/src/Hal/Serializer/CollectionNormalizer.php#L86-L100

The changes in `ApiTestCaseTest` (testing `jsonld` and `jsonhal` as formats) trigger a failure condition without this change:
```
1) ApiPlatform\Tests\Symfony\Bundle\Test\ApiTestCaseTest::testAssertMatchesResourceCollectionJsonSchema with data set "jsonhal" ('jsonhal', 'application/hal+json')
Failed asserting that Array &0 (
    '_links' => Array &1 (
        'self' => Array &2 (
            'href' => '/resource_interfaces'
        )
        'item' => Array &3 (
            0 => Array &4 (
                'href' => '/resource_interfaces/item1'
            )
            1 => Array &5 (
                'href' => '/resource_interfaces/item2'
            )
        )
    )
    '_embedded' => Array &6 (
        'item' => Array &7 (
            0 => Array &8 (
                '_links' => Array &9 (
                    'self' => Array &10 (
                        'href' => '/resource_interfaces/item1'
                    )
                )
                'foo' => 'item1'
                'fooz' => 'fooz'
            )
            1 => Array &11 (
                '_links' => Array &12 (
                    'self' => Array &13 (
                        'href' => '/resource_interfaces/item2'
                    )
                )
                'foo' => 'item2'
                'fooz' => 'fooz'
            )
        )
    )
) matches the provided JSON Schema.
_embedded: Object value found, but an array is required

2) ApiPlatform\Tests\Symfony\Bundle\Test\ApiTestCaseTest::testAssertMatchesResourceCollectionJsonSchemaKeepSerializationContext with data set "jsonhal" ('jsonhal', 'application/hal+json')
Failed asserting that Array &0 (
    '_links' => Array &1 (
        'self' => Array &2 (
            'href' => '/issue-6146-parents'
        )
        'item' => Array &3 (
            0 => Array &4 (
                'href' => '/issue-6146-parents/1'
            )
        )
    )
    'totalItems' => 1
    'itemsPerPage' => 3
    '_embedded' => Array &5 (
        'item' => Array &6 (
            0 => Array &7 (
                '_links' => Array &8 (
                    'self' => Array &9 (
                        'href' => '/issue-6146-parents/1'
                    )
                )
            )
        )
    )
) matches the provided JSON Schema.
_embedded: Object value found, but an array is required

3) ApiPlatform\Tests\Symfony\Bundle\Test\ApiTestCaseTest::testAssertMatchesResourceItemAndCollectionJsonSchemaOutputWithContext with data set "jsonhal" ('jsonhal', 'application/hal+json')
Failed asserting that Array &0 (
    '_links' => Array &1 (
        'self' => Array &2 (
            'href' => '/users-with-groups'
        )
        'item' => Array &3 (
            0 => Array &4 (
                'href' => '/users/1'
            )
        )
    )
    'totalItems' => 1
    'itemsPerPage' => 3
    '_embedded' => Array &5 (
        'item' => Array &6 (
            0 => Array &7 (
                '_links' => Array &8 (
                    'self' => Array &9 (
                        'href' => '/users/1'
                    )
                )
                'fullname' => 'Grégoire'
            )
        )
    )
) matches the provided JSON Schema.
_embedded: Object value found, but an array is required
```

Reply:
From https://datatracker.ietf.org/doc/html/draft-kelly-json-hal#section-6

```json
{
  "_links": {
    "self": { "href": "/orders" },
    "next": { "href": "/orders?page=2" },
    "find": { "href": "/orders{?id}", "templated": true }
  },
  "_embedded": {
    "orders": [{
        "_links": {
          "self": { "href": "/orders/123" },
          "basket": { "href": "/baskets/98712" },
          "customer": { "href": "/customers/7809" }
        },
        "total": 30.00,
        "currency": "USD",
        "status": "shipped",
      },{
        "_links": {
          "self": { "href": "/orders/124" },
          "basket": { "href": "/baskets/97213" },
          "customer": { "href": "/customers/12369" }
        },
        "total": 20.00,
        "currency": "USD",
        "status": "processing"
    }]
  },
  "currentlyProcessing": 14,
  "shippedToday": 20
}
```

Which also comes back to a previous question as to the key(s) of `_embedded` … we use `items` but the examples in the draft seem to use the resource name.

Reply:
To be correct the schema should be `anyOf` object or array. We force `item` at https://github.com/api-platform/core/blob/874e4d6707ecdd48b27ed91b0b66ea5e33691e7c/src/Hal/Serializer/CollectionNormalizer.php#L95 therefore we need the key at `item` as I see it.

Reply:
As in how I did it originally in https://github.com/api-platform/core/commit/97f7dff15db98f7cd0e8510fc828557f35513397?

Reply:
Ugh, never mind, it clicked what you meant … now I (think) I get you :innocent:

Reply:
Looking too hard at the same thing for too long and got my brain in a twist.

Reply:
Right now we may have (as this is what we use for our collections):

```
'anyOf' => [
    [
        'type' => 'object',
        'properties' => [
            'item' => [
                'type' => 'array',
                'items' => $items,
            ],
        ],
    ],
    [ 'type' => 'object' ]
],
```

Although the spec says that it can also be

```
  [
        'type' => 'array'
  ],
```

I'm okay with adding it if you want